File: generated_code/python/test_unittest_203.py
import os
import importlib.util
import unittest

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise ImportError("TEST_SOLUTION_FILE environment variable is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution
ListNode = sol_module.ListNode


class TestRemoveLinkedListElements(unittest.TestCase):

    def _create_linked_list(self, arr):
        """Helper method to create a linked list from a list."""
        if not arr:
            return None
        head = ListNode(arr[0])
        current = head
        for val in arr[1:]:
            current.next = ListNode(val)
            current = current.next
        return head

    def _linked_list_to_list(self, head):
        """Helper method to convert a linked list back to a Python list."""
        result = []
        current = head
        while current:
            result.append(current.val)
            current = current.next
        return result

    def test_example_1_mixed_elements(self):
        """Test Example 1: Standard case with target elements in middle and end."""
        head = self._create_linked_list([1, 2, 6, 3, 4, 5, 6])
        val = 6
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [1, 2, 3, 4, 5])

    def test_example_2_empty_list(self):
        """Test Example 2: Edge case where the input list is completely empty."""
        head = self._create_linked_list([])
        val = 1
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [])

    def test_example_3_all_elements_match(self):
        """Test Example 3: Case where every single node matches the target value."""
        head = self._create_linked_list([7, 7, 7, 7])
        val = 7
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [])

    def test_edge_case_single_element_matching(self):
        """Test edge case with a single node list that matches the target value."""
        head = self._create_linked_list([5])
        val = 5
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [])

    def test_edge_case_single_element_not_matching(self):
        """Test edge case with a single node list that does not match the target value."""
        head = self._create_linked_list([5])
        val = 3
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [5])

    def test_target_at_head_only(self):
        """Test case where target elements are strictly at the beginning of the list."""
        head = self._create_linked_list([1, 1, 2, 3])
        val = 1
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [2, 3])

    def test_target_at_tail_only(self):
        """Test case where target elements are strictly at the end of the list."""
        head = self._create_linked_list([2, 3, 4, 4])
        val = 4
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [2, 3])

    def test_no_elements_match(self):
        """Test case where no elements in the list match the target value."""
        head = self._create_linked_list([1, 2, 3, 4, 5])
        val = 0
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [1, 2, 3, 4, 5])

    def test_consecutive_target_elements_interspersed(self):
        """Test case containing multiple consecutive target values interspersed with non-target values."""
        head = self._create_linked_list([1, 2, 2, 3, 2, 2, 2, 4, 2])
        val = 2
        sol = Solution()
        new_head = sol.removeElements(head, val)
        self.assertEqual(self._linked_list_to_list(new_head), [1, 3, 4])


if __name__ == "__main__":
    unittest.main()

File: generated_code/cpp/test_unittest_203.cpp
#include <gtest/gtest.h>
#include <vector>
#include "solution_proxy.h"

// Helper function to create a linked list from a vector
ListNode* createLinkedList(const std::vector<int>& values) {
    if (values.empty()) return nullptr;
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    for (size_t i = 1; i < values.size(); ++i) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    return head;
}

// Helper function to convert a linked list back to a vector
std::vector<int> linkedListToVector(ListNode* head) {
    std::vector<int> result;
    ListNode* current = head;
    while (current != nullptr) {
        result.push_back(current->val);
        current = current->next;
    }
    return result;
}

// Helper function to free the memory of a linked list
void freeLinkedList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

// Test Class Setup
class RemoveElementsTest : public ::testing::Test {
protected:
    Solution solution;
};

// Example 1: Standard case with some target elements in the middle and end
TEST_F(RemoveElementsTest, Example1_MixedElements) {
    ListNode* head = createLinkedList({1, 2, 6, 3, 4, 5, 6});
    int val = 6;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 2, 3, 4, 5};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Example 2: Empty list input
TEST_F(RemoveElementsTest, Example2_EmptyList) {
    ListNode* head = createLinkedList({});
    int val = 1;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Example 3: All elements in the list match the target value
TEST_F(RemoveElementsTest, Example3_AllElementsMatch) {
    ListNode* head = createLinkedList({7, 7, 7, 7});
    int val = 7;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Target value appears at the head of the list
TEST_F(RemoveElementsTest, TargetAtHeadOnly) {
    ListNode* head = createLinkedList({5, 1, 2, 3});
    int val = 5;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 2, 3};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Target value appears consecutively at the start
TEST_F(RemoveElementsTest, ConsecutiveTargetsAtHead) {
    ListNode* head = createLinkedList({5, 5, 5, 1, 2});
    int val = 5;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 2};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Target value appears consecutively in the middle
TEST_F(RemoveElementsTest, ConsecutiveTargetsInMiddle) {
    ListNode* head = createLinkedList({1, 5, 5, 5, 2});
    int val = 5;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 2};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Target value is not present in the list
TEST_F(RemoveElementsTest, TargetValueNotFound) {
    ListNode* head = createLinkedList({1, 2, 3, 4});
    int val = 9;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 2, 3, 4};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Single element list matching the target value
TEST_F(RemoveElementsTest, SingleElementMatches) {
    ListNode* head = createLinkedList({42});
    int val = 42;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Edge Case: Single element list not matching the target value
TEST_F(RemoveElementsTest, SingleElementDoesNotMatch) {
    ListNode* head = createLinkedList({42});
    int val = 24;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {42};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}

// Constraint Test: Node values and val parameter edge combinations (min/max boundaries)
TEST_F(RemoveElementsTest, BoundaryValueConstraints) {
    ListNode* head = createLinkedList({1, 50, 1, 50});
    int val = 50;
    
    ListNode* resultHead = solution.removeElements(head, val);
    std::vector<int> resultVec = linkedListToVector(resultHead);
    std::vector<int> expected = {1, 1};
    
    EXPECT_EQ(resultVec, expected);
    freeLinkedList(resultHead);
}